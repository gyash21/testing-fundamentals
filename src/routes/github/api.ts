import type { paths } from "@octokit/openapi-types";

type OrgRepoResponse =
  paths["/repos/{owner}/{repo}"]["get"]["responses"]["200"]["content"]["application/json"];
type UserReposResponse =
  paths["/users/{username}/repos"]["get"]["responses"]["200"]["content"]["application/json"];
export type UserRepository = UserReposResponse[0];

export type Fetch = typeof fetch;

export function delay(time: number) {
  return new Promise<void>((resolve) => setTimeout(resolve, time));
}

export class GitHubAPI {
  private fetch: Fetch;
  private accessToken?: string;
  private delay: (time: number) => Promise<void>;

  constructor(
    fetch: Fetch,
    accessToken?: string,
    _delay: (time: number) => Promise<void> = delay
  ) {
    this.fetch = fetch;
    this.accessToken = accessToken;
    this.delay = _delay;
  }

  async getRepository(user: string, repo: string) {
    const headers = this.createHeaders();

    return Promise.race([
      this.fetchJSON<OrgRepoResponse>(
        `https://api.github.com/repos/${user}/${repo}`,
        headers
      ),
      this.delay(5000).then(() => ({ response: "TimeOut" })),
    ]);
  }

  async getRepositories(user: string) {
    const headers = this.createHeaders();
    const perPage = 30;
    let page = 1;
    let repositories: UserRepository[] = [];
    let hasMore = true;
    while (hasMore) {
      const repos = await this.fetchJSON<UserReposResponse>(
        `https://api.github.com/users/${user}/repos?per_page=${perPage}&page=${page++}`,
        headers
      );
      repositories.push(...repos);
      hasMore = repos.length === perPage;
    }
    return repositories;
  }

  getMostPopularRepository(user: string) {
    return this.getRepositories(user).then((repos) =>
      repos.reduce((prev, curr) =>
        (prev.stargazers_count || 0) > (curr.stargazers_count || 0)
          ? prev
          : curr
      )
    );
  }

  private createHeaders() {
    const headers: HeadersInit = {
      "User-Agent": "Qwik Workshop",
      "X-GitHub-Api-Version": "2022-11-28",
    };
    if (this.accessToken) {
      headers["Authorization"] = "Bearer " + this.accessToken;
    }
    return headers;
  }

  private async fetchJSON<T>(url: string, headers: HeadersInit) {
    const x = this.fetch(url, { headers });
    const response = await x;
    const repository = (await response.json()) as T;
    return repository;
  }
}
