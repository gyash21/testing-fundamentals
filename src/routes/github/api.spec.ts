import { describe, it, vi, Mock, beforeEach } from "vitest";
import { Fetch, GitHubAPI, delay } from "./api";

describe("GithubAPI", () => {
  describe("repository", () => {
    let fetchMock: Mock<Parameters<Fetch>, Promise<Response>>;
    let delayMock: Mock<[number], Promise<void>>;
    let api: GitHubAPI;
    beforeEach(() => {
      fetchMock = vi.fn<Parameters<Fetch>, Promise<Response>>(mockPromise);
      delayMock = vi.fn<[number], Promise<void>>(mockPromise);
      api = new GitHubAPI(fetchMock, "TEST_ACCESS_TOKEN", delayMock);
    });

    it("should repository details", async ({ expect }) => {
      const repoPromise = api.getRepository("mhevery", "qwik");
      expect(fetchMock).toHaveBeenCalledWith(
        "https://api.github.com/repos/mhevery/qwik",
        {
          headers: {
            "User-Agent": "Qwik Workshop",
            "X-GitHub-Api-Version": "2022-11-28",
            Authorization: "Bearer TEST_ACCESS_TOKEN",
          },
        }
      );
      fetchMock.mock.results[0].value.resolve!(new Response('"MockResponse"'));
      expect(await repoPromise).toEqual("MockResponse");
    });

    it("should wait for some time before giving up", async ({ expect }) => {
      const repoPromise = api.getRepository("mhevery", "qwik");
      expect(fetchMock).toHaveBeenCalledWith(
        "https://api.github.com/repos/mhevery/qwik",
        expect.any(Object)
      );
      expect(delayMock).toHaveBeenCalledWith(5000);
      delayMock.mock.results[0].value.resolve!();
      await delay(0);
      fetchMock.mock.results[0].value.resolve!();
      expect(await repoPromise).toEqual({ response: "TimeOut" });
    });

    it("should resolve even if before timeout", async ({ expect }) => {
      const repoPromise = api.getRepository("mhevery", "qwik");
      expect(fetchMock).toHaveBeenCalledWith(
        "https://api.github.com/repos/mhevery/qwik",
        expect.any(Object)
      );
      expect(delayMock).toHaveBeenCalledWith(5000);
      fetchMock.mock.results[0].value.resolve!(new Response('"MockResponse"'));
      await delay(0);
      delayMock.mock.results[0].value.resolve!();
      expect(await repoPromise).toEqual("MockResponse");
    });
    describe("getRepositories", () => {
      it("should fetch repositories of a user", async ({ expect }) => {
        const repoPromise = api.getRepositories("mhevery");
        expect(fetchMock).toHaveBeenCalledWith(
          "https://api.github.com/users/mhevery/repos?per_page=30&page=1",
          expect.any(Object)
        );
        fetchMock.mock.results[0].value.resolve!(
          new Response('["MockResponse"]')
        );
        expect(await repoPromise).toEqual(["MockResponse"]);
      });
      it("should paginate to get all", async ({ expect }) => {
        const repoPromise = api.getRepositories("mhevery");
        expect(fetchMock).toHaveBeenCalledWith(
          "https://api.github.com/users/mhevery/repos?per_page=30&page=1",
          expect.any(Object)
        );
        fetchMock.mock.results[0].value.resolve!(
          new Response('["MockResponse"]')
        );
        expect(await repoPromise).toEqual(["MockResponse"]);
      });
    });

    it("should get most popular repository of a user", async ({ expect }) => {
      const repoPromise = api.getMostPopularRepository("mhevery");
      expect(fetchMock).toHaveBeenCalledWith(
        "https://api.github.com/users/mhevery/repos?per_page=30&page=1",
        expect.any(Object)
      );
      fetchMock.mock.results[0].value.resolve!(
        new Response(
          JSON.stringify([
            { id: 1, stargazers_count: 100 },
            { id: 2, stargazers_count: 200 },
          ])
        )
      );
      expect(await repoPromise).toEqual({ id: 2, stargazers_count: 200 });
    });
  });
});

function mockPromise<T>(): Promise<T> & {
  resolve: (value: T) => void;
  reject: (error: any) => void;
} {
  let resolve!: (value: T) => void;
  let reject!: (error: any) => void;
  const promise = new Promise<T>((res, rej) => {
    resolve = res;
    reject = rej;
  }) as any;
  promise.resolve = resolve;
  promise.reject = reject;
  return promise;
}
