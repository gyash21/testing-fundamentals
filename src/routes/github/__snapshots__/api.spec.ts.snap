// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GithubAPI > repository > should repository details 1`] = `
{
  "allow_forking": true,
  "archive_url": "https://api.github.com/repos/mhevery/qwik/{archive_format}{/ref}",
  "archived": false,
  "assignees_url": "https://api.github.com/repos/mhevery/qwik/assignees{/user}",
  "blobs_url": "https://api.github.com/repos/mhevery/qwik/git/blobs{/sha}",
  "branches_url": "https://api.github.com/repos/mhevery/qwik/branches{/branch}",
  "clone_url": "https://github.com/mhevery/qwik.git",
  "collaborators_url": "https://api.github.com/repos/mhevery/qwik/collaborators{/collaborator}",
  "comments_url": "https://api.github.com/repos/mhevery/qwik/comments{/number}",
  "commits_url": "https://api.github.com/repos/mhevery/qwik/commits{/sha}",
  "compare_url": "https://api.github.com/repos/mhevery/qwik/compare/{base}...{head}",
  "contents_url": "https://api.github.com/repos/mhevery/qwik/contents/{+path}",
  "contributors_url": "https://api.github.com/repos/mhevery/qwik/contributors",
  "created_at": "2021-06-01T21:33:00Z",
  "default_branch": "main",
  "deployments_url": "https://api.github.com/repos/mhevery/qwik/deployments",
  "description": "An Open-Source framework designed for best possible [time to interactive](https://web.dev/interactive/), by focusing on [resumability](./docs/RESUMABLE.md) of server-side-rendering of HTML, and [fine-grained lazy-loading](./docs/LAZY_LOADING.md) of code.",
  "disabled": false,
  "downloads_url": "https://api.github.com/repos/mhevery/qwik/downloads",
  "events_url": "https://api.github.com/repos/mhevery/qwik/events",
  "fork": true,
  "forks": 1,
  "forks_count": 1,
  "forks_url": "https://api.github.com/repos/mhevery/qwik/forks",
  "full_name": "mhevery/qwik",
  "git_commits_url": "https://api.github.com/repos/mhevery/qwik/git/commits{/sha}",
  "git_refs_url": "https://api.github.com/repos/mhevery/qwik/git/refs{/sha}",
  "git_tags_url": "https://api.github.com/repos/mhevery/qwik/git/tags{/sha}",
  "git_url": "git://github.com/mhevery/qwik.git",
  "has_discussions": false,
  "has_downloads": true,
  "has_issues": false,
  "has_pages": false,
  "has_projects": true,
  "has_wiki": true,
  "homepage": null,
  "hooks_url": "https://api.github.com/repos/mhevery/qwik/hooks",
  "html_url": "https://github.com/mhevery/qwik",
  "id": 372966483,
  "is_template": false,
  "issue_comment_url": "https://api.github.com/repos/mhevery/qwik/issues/comments{/number}",
  "issue_events_url": "https://api.github.com/repos/mhevery/qwik/issues/events{/number}",
  "issues_url": "https://api.github.com/repos/mhevery/qwik/issues{/number}",
  "keys_url": "https://api.github.com/repos/mhevery/qwik/keys{/key_id}",
  "labels_url": "https://api.github.com/repos/mhevery/qwik/labels{/name}",
  "language": "TypeScript",
  "languages_url": "https://api.github.com/repos/mhevery/qwik/languages",
  "license": {
    "key": "mit",
    "name": "MIT License",
    "node_id": "MDc6TGljZW5zZTEz",
    "spdx_id": "MIT",
    "url": "https://api.github.com/licenses/mit",
  },
  "merges_url": "https://api.github.com/repos/mhevery/qwik/merges",
  "milestones_url": "https://api.github.com/repos/mhevery/qwik/milestones{/number}",
  "mirror_url": null,
  "name": "qwik",
  "network_count": 1187,
  "node_id": "MDEwOlJlcG9zaXRvcnkzNzI5NjY0ODM=",
  "notifications_url": "https://api.github.com/repos/mhevery/qwik/notifications{?since,all,participating}",
  "open_issues": 0,
  "open_issues_count": 0,
  "owner": {
    "avatar_url": "https://avatars.githubusercontent.com/u/111951?v=4",
    "events_url": "https://api.github.com/users/mhevery/events{/privacy}",
    "followers_url": "https://api.github.com/users/mhevery/followers",
    "following_url": "https://api.github.com/users/mhevery/following{/other_user}",
    "gists_url": "https://api.github.com/users/mhevery/gists{/gist_id}",
    "gravatar_id": "",
    "html_url": "https://github.com/mhevery",
    "id": 111951,
    "login": "mhevery",
    "node_id": "MDQ6VXNlcjExMTk1MQ==",
    "organizations_url": "https://api.github.com/users/mhevery/orgs",
    "received_events_url": "https://api.github.com/users/mhevery/received_events",
    "repos_url": "https://api.github.com/users/mhevery/repos",
    "site_admin": false,
    "starred_url": "https://api.github.com/users/mhevery/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/mhevery/subscriptions",
    "type": "User",
    "url": "https://api.github.com/users/mhevery",
  },
  "parent": {
    "allow_forking": true,
    "archive_url": "https://api.github.com/repos/BuilderIO/qwik/{archive_format}{/ref}",
    "archived": false,
    "assignees_url": "https://api.github.com/repos/BuilderIO/qwik/assignees{/user}",
    "blobs_url": "https://api.github.com/repos/BuilderIO/qwik/git/blobs{/sha}",
    "branches_url": "https://api.github.com/repos/BuilderIO/qwik/branches{/branch}",
    "clone_url": "https://github.com/BuilderIO/qwik.git",
    "collaborators_url": "https://api.github.com/repos/BuilderIO/qwik/collaborators{/collaborator}",
    "comments_url": "https://api.github.com/repos/BuilderIO/qwik/comments{/number}",
    "commits_url": "https://api.github.com/repos/BuilderIO/qwik/commits{/sha}",
    "compare_url": "https://api.github.com/repos/BuilderIO/qwik/compare/{base}...{head}",
    "contents_url": "https://api.github.com/repos/BuilderIO/qwik/contents/{+path}",
    "contributors_url": "https://api.github.com/repos/BuilderIO/qwik/contributors",
    "created_at": "2021-05-19T15:33:32Z",
    "default_branch": "main",
    "deployments_url": "https://api.github.com/repos/BuilderIO/qwik/deployments",
    "description": "Instant-loading web apps, without effort",
    "disabled": false,
    "downloads_url": "https://api.github.com/repos/BuilderIO/qwik/downloads",
    "events_url": "https://api.github.com/repos/BuilderIO/qwik/events",
    "fork": false,
    "forks": 1187,
    "forks_count": 1187,
    "forks_url": "https://api.github.com/repos/BuilderIO/qwik/forks",
    "full_name": "BuilderIO/qwik",
    "git_commits_url": "https://api.github.com/repos/BuilderIO/qwik/git/commits{/sha}",
    "git_refs_url": "https://api.github.com/repos/BuilderIO/qwik/git/refs{/sha}",
    "git_tags_url": "https://api.github.com/repos/BuilderIO/qwik/git/tags{/sha}",
    "git_url": "git://github.com/BuilderIO/qwik.git",
    "has_discussions": false,
    "has_downloads": true,
    "has_issues": true,
    "has_pages": false,
    "has_projects": true,
    "has_wiki": false,
    "homepage": "https://qwik.dev",
    "hooks_url": "https://api.github.com/repos/BuilderIO/qwik/hooks",
    "html_url": "https://github.com/BuilderIO/qwik",
    "id": 368917877,
    "is_template": false,
    "issue_comment_url": "https://api.github.com/repos/BuilderIO/qwik/issues/comments{/number}",
    "issue_events_url": "https://api.github.com/repos/BuilderIO/qwik/issues/events{/number}",
    "issues_url": "https://api.github.com/repos/BuilderIO/qwik/issues{/number}",
    "keys_url": "https://api.github.com/repos/BuilderIO/qwik/keys{/key_id}",
    "labels_url": "https://api.github.com/repos/BuilderIO/qwik/labels{/name}",
    "language": "TypeScript",
    "languages_url": "https://api.github.com/repos/BuilderIO/qwik/languages",
    "license": {
      "key": "mit",
      "name": "MIT License",
      "node_id": "MDc6TGljZW5zZTEz",
      "spdx_id": "MIT",
      "url": "https://api.github.com/licenses/mit",
    },
    "merges_url": "https://api.github.com/repos/BuilderIO/qwik/merges",
    "milestones_url": "https://api.github.com/repos/BuilderIO/qwik/milestones{/number}",
    "mirror_url": null,
    "name": "qwik",
    "node_id": "MDEwOlJlcG9zaXRvcnkzNjg5MTc4Nzc=",
    "notifications_url": "https://api.github.com/repos/BuilderIO/qwik/notifications{?since,all,participating}",
    "open_issues": 396,
    "open_issues_count": 396,
    "owner": {
      "avatar_url": "https://avatars.githubusercontent.com/u/35700027?v=4",
      "events_url": "https://api.github.com/users/BuilderIO/events{/privacy}",
      "followers_url": "https://api.github.com/users/BuilderIO/followers",
      "following_url": "https://api.github.com/users/BuilderIO/following{/other_user}",
      "gists_url": "https://api.github.com/users/BuilderIO/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/BuilderIO",
      "id": 35700027,
      "login": "BuilderIO",
      "node_id": "MDEyOk9yZ2FuaXphdGlvbjM1NzAwMDI3",
      "organizations_url": "https://api.github.com/users/BuilderIO/orgs",
      "received_events_url": "https://api.github.com/users/BuilderIO/received_events",
      "repos_url": "https://api.github.com/users/BuilderIO/repos",
      "site_admin": false,
      "starred_url": "https://api.github.com/users/BuilderIO/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/BuilderIO/subscriptions",
      "type": "Organization",
      "url": "https://api.github.com/users/BuilderIO",
    },
    "private": false,
    "pulls_url": "https://api.github.com/repos/BuilderIO/qwik/pulls{/number}",
    "pushed_at": "2024-03-01T02:32:49Z",
    "releases_url": "https://api.github.com/repos/BuilderIO/qwik/releases{/id}",
    "size": 37500,
    "ssh_url": "git@github.com:BuilderIO/qwik.git",
    "stargazers_count": 19909,
    "stargazers_url": "https://api.github.com/repos/BuilderIO/qwik/stargazers",
    "statuses_url": "https://api.github.com/repos/BuilderIO/qwik/statuses/{sha}",
    "subscribers_url": "https://api.github.com/repos/BuilderIO/qwik/subscribers",
    "subscription_url": "https://api.github.com/repos/BuilderIO/qwik/subscription",
    "svn_url": "https://github.com/BuilderIO/qwik",
    "tags_url": "https://api.github.com/repos/BuilderIO/qwik/tags",
    "teams_url": "https://api.github.com/repos/BuilderIO/qwik/teams",
    "topics": [
      "framework",
      "javascript",
      "web",
    ],
    "trees_url": "https://api.github.com/repos/BuilderIO/qwik/git/trees{/sha}",
    "updated_at": "2024-03-01T00:53:17Z",
    "url": "https://api.github.com/repos/BuilderIO/qwik",
    "visibility": "public",
    "watchers": 19909,
    "watchers_count": 19909,
    "web_commit_signoff_required": false,
  },
  "private": false,
  "pulls_url": "https://api.github.com/repos/mhevery/qwik/pulls{/number}",
  "pushed_at": "2024-02-06T02:58:07Z",
  "releases_url": "https://api.github.com/repos/mhevery/qwik/releases{/id}",
  "size": 35608,
  "source": {
    "allow_forking": true,
    "archive_url": "https://api.github.com/repos/BuilderIO/qwik/{archive_format}{/ref}",
    "archived": false,
    "assignees_url": "https://api.github.com/repos/BuilderIO/qwik/assignees{/user}",
    "blobs_url": "https://api.github.com/repos/BuilderIO/qwik/git/blobs{/sha}",
    "branches_url": "https://api.github.com/repos/BuilderIO/qwik/branches{/branch}",
    "clone_url": "https://github.com/BuilderIO/qwik.git",
    "collaborators_url": "https://api.github.com/repos/BuilderIO/qwik/collaborators{/collaborator}",
    "comments_url": "https://api.github.com/repos/BuilderIO/qwik/comments{/number}",
    "commits_url": "https://api.github.com/repos/BuilderIO/qwik/commits{/sha}",
    "compare_url": "https://api.github.com/repos/BuilderIO/qwik/compare/{base}...{head}",
    "contents_url": "https://api.github.com/repos/BuilderIO/qwik/contents/{+path}",
    "contributors_url": "https://api.github.com/repos/BuilderIO/qwik/contributors",
    "created_at": "2021-05-19T15:33:32Z",
    "default_branch": "main",
    "deployments_url": "https://api.github.com/repos/BuilderIO/qwik/deployments",
    "description": "Instant-loading web apps, without effort",
    "disabled": false,
    "downloads_url": "https://api.github.com/repos/BuilderIO/qwik/downloads",
    "events_url": "https://api.github.com/repos/BuilderIO/qwik/events",
    "fork": false,
    "forks": 1187,
    "forks_count": 1187,
    "forks_url": "https://api.github.com/repos/BuilderIO/qwik/forks",
    "full_name": "BuilderIO/qwik",
    "git_commits_url": "https://api.github.com/repos/BuilderIO/qwik/git/commits{/sha}",
    "git_refs_url": "https://api.github.com/repos/BuilderIO/qwik/git/refs{/sha}",
    "git_tags_url": "https://api.github.com/repos/BuilderIO/qwik/git/tags{/sha}",
    "git_url": "git://github.com/BuilderIO/qwik.git",
    "has_discussions": false,
    "has_downloads": true,
    "has_issues": true,
    "has_pages": false,
    "has_projects": true,
    "has_wiki": false,
    "homepage": "https://qwik.dev",
    "hooks_url": "https://api.github.com/repos/BuilderIO/qwik/hooks",
    "html_url": "https://github.com/BuilderIO/qwik",
    "id": 368917877,
    "is_template": false,
    "issue_comment_url": "https://api.github.com/repos/BuilderIO/qwik/issues/comments{/number}",
    "issue_events_url": "https://api.github.com/repos/BuilderIO/qwik/issues/events{/number}",
    "issues_url": "https://api.github.com/repos/BuilderIO/qwik/issues{/number}",
    "keys_url": "https://api.github.com/repos/BuilderIO/qwik/keys{/key_id}",
    "labels_url": "https://api.github.com/repos/BuilderIO/qwik/labels{/name}",
    "language": "TypeScript",
    "languages_url": "https://api.github.com/repos/BuilderIO/qwik/languages",
    "license": {
      "key": "mit",
      "name": "MIT License",
      "node_id": "MDc6TGljZW5zZTEz",
      "spdx_id": "MIT",
      "url": "https://api.github.com/licenses/mit",
    },
    "merges_url": "https://api.github.com/repos/BuilderIO/qwik/merges",
    "milestones_url": "https://api.github.com/repos/BuilderIO/qwik/milestones{/number}",
    "mirror_url": null,
    "name": "qwik",
    "node_id": "MDEwOlJlcG9zaXRvcnkzNjg5MTc4Nzc=",
    "notifications_url": "https://api.github.com/repos/BuilderIO/qwik/notifications{?since,all,participating}",
    "open_issues": 396,
    "open_issues_count": 396,
    "owner": {
      "avatar_url": "https://avatars.githubusercontent.com/u/35700027?v=4",
      "events_url": "https://api.github.com/users/BuilderIO/events{/privacy}",
      "followers_url": "https://api.github.com/users/BuilderIO/followers",
      "following_url": "https://api.github.com/users/BuilderIO/following{/other_user}",
      "gists_url": "https://api.github.com/users/BuilderIO/gists{/gist_id}",
      "gravatar_id": "",
      "html_url": "https://github.com/BuilderIO",
      "id": 35700027,
      "login": "BuilderIO",
      "node_id": "MDEyOk9yZ2FuaXphdGlvbjM1NzAwMDI3",
      "organizations_url": "https://api.github.com/users/BuilderIO/orgs",
      "received_events_url": "https://api.github.com/users/BuilderIO/received_events",
      "repos_url": "https://api.github.com/users/BuilderIO/repos",
      "site_admin": false,
      "starred_url": "https://api.github.com/users/BuilderIO/starred{/owner}{/repo}",
      "subscriptions_url": "https://api.github.com/users/BuilderIO/subscriptions",
      "type": "Organization",
      "url": "https://api.github.com/users/BuilderIO",
    },
    "private": false,
    "pulls_url": "https://api.github.com/repos/BuilderIO/qwik/pulls{/number}",
    "pushed_at": "2024-03-01T02:32:49Z",
    "releases_url": "https://api.github.com/repos/BuilderIO/qwik/releases{/id}",
    "size": 37500,
    "ssh_url": "git@github.com:BuilderIO/qwik.git",
    "stargazers_count": 19909,
    "stargazers_url": "https://api.github.com/repos/BuilderIO/qwik/stargazers",
    "statuses_url": "https://api.github.com/repos/BuilderIO/qwik/statuses/{sha}",
    "subscribers_url": "https://api.github.com/repos/BuilderIO/qwik/subscribers",
    "subscription_url": "https://api.github.com/repos/BuilderIO/qwik/subscription",
    "svn_url": "https://github.com/BuilderIO/qwik",
    "tags_url": "https://api.github.com/repos/BuilderIO/qwik/tags",
    "teams_url": "https://api.github.com/repos/BuilderIO/qwik/teams",
    "topics": [
      "framework",
      "javascript",
      "web",
    ],
    "trees_url": "https://api.github.com/repos/BuilderIO/qwik/git/trees{/sha}",
    "updated_at": "2024-03-01T00:53:17Z",
    "url": "https://api.github.com/repos/BuilderIO/qwik",
    "visibility": "public",
    "watchers": 19909,
    "watchers_count": 19909,
    "web_commit_signoff_required": false,
  },
  "ssh_url": "git@github.com:mhevery/qwik.git",
  "stargazers_count": 8,
  "stargazers_url": "https://api.github.com/repos/mhevery/qwik/stargazers",
  "statuses_url": "https://api.github.com/repos/mhevery/qwik/statuses/{sha}",
  "subscribers_count": 2,
  "subscribers_url": "https://api.github.com/repos/mhevery/qwik/subscribers",
  "subscription_url": "https://api.github.com/repos/mhevery/qwik/subscription",
  "svn_url": "https://github.com/mhevery/qwik",
  "tags_url": "https://api.github.com/repos/mhevery/qwik/tags",
  "teams_url": "https://api.github.com/repos/mhevery/qwik/teams",
  "temp_clone_token": null,
  "topics": [],
  "trees_url": "https://api.github.com/repos/mhevery/qwik/git/trees{/sha}",
  "updated_at": "2024-02-06T10:47:40Z",
  "url": "https://api.github.com/repos/mhevery/qwik",
  "visibility": "public",
  "watchers": 8,
  "watchers_count": 8,
  "web_commit_signoff_required": false,
}
`;
